lib/utils/selector.js

```diff
@@ -0,0 +1,582 @@
+        result ||
+          // for :has()
+          ((element, subject) => element === subject),
+        combinator,
+        right
+      )
+    } else {
+      const sel = nodeToVElementMatcher(node)
+      result = result ? compound(result, sel) : sel
+    }
+    node = nodes.shift()
+  }
+  if (!result) {
+    throw new SelectorError(`Unexpected empty selector.`)
+  }
+  return result
+}
+
+/**
+ * @param {VElementMatcher} left
+ * @param {string} combinator
+ * @param {VElementMatcher} right
+ * @returns {VElementMatcher}
+ */
+function combination(left, combinator, right) {
+  if (!combinator.trim()) {
+    // descendant
+    return (element, subject) => {
+      if (right(element, null)) {
+        let parent = element.parent
+        while (parent.type === 'VElement') {
+          if (left(parent, subject)) {
+            return true
+          }
+          parent = parent.parent
+        }
+      }
+      return false
+    }
+  }
+
+  if (combinator === '>') {
+    // child
+    return (element, subject) => {
+      if (right(element, null)) {
+        const parent = element.parent
+        if (parent.type === 'VElement') {
+          return left(parent, subject)
+        }
+      }
+      return false
+    }
+  }
+
+  if (combinator === '+') {
+    // adjacent
+    return (element, subject) => {
+      if (right(element, null)) {
+        const before = getBeforeElement(element)
+        if (before) {
+          return left(before, subject)
+        }
+      }
+      return false
+    }
+  }
+
+  if (combinator === '~') {
+    // sibling
+    return (element, subject) => {
+      if (right(element, null)) {
+        for (const before of getBeforeElements(element)) {
+          if (left(before, subject)) {
+            return true
+          }
+        }
+      }
+      return false
+    }
+  }
+  throw new SelectorError(`Unknown combinator: ${combinator}.`)
+}
+
+/**
+ * Convert node to VElementMatcher
+ * @param {Exclude<parser.Node, {type:'combinator'|'comment'|'root'|'selector'}>} selector
+ * @returns {VElementMatcher}
+ */
+function nodeToVElementMatcher(selector) {
+  if (selector.type === 'attribute') {
+    return attributeNodeToVElementMatcher(selector)
+  }
+  if (selector.type === 'class') {
+    const className = selector.value
+    return (element) => {
+      const attrValue = getAttributeValue(element, 'class')
+      if (attrValue == null) {
+        return false
+      }
+      return attrValue.split(/\s+/gu).includes(className)
+    }
+  }
+  if (selector.type === 'id') {
+    const id = selector.value
+    return (element) => {
+      const attrValue = getAttributeValue(element, 'id')
+      if (attrValue == null) {
+        return false
+      }
+      return attrValue === id
+    }
+  }
+  if (selector.type === 'tag') {
+    const name = selector.value
+    return (element) => element.rawName === name
+  }
+  if (selector.type === 'universal') {
+    return () => true
+  }
+
+  if (selector.type === 'pseudo') {
```

Comment:
Maybe extract everything inside this branch into a separate function `pseudoSelectorToVelementMatcher` or similar?